python/ray/scripts/symmetric_run.py

```diff
@@ -0,0 +1,259 @@
+"""Symmetric Run for Ray."""
+
+from typing import List
+
+import click
+import ray
+import socket
+import psutil
+import subprocess
+import sys
+import time
+
+from ray._private.ray_constants import env_integer
+from ray._raylet import GcsClient
+
+CLUSTER_WAIT_TIMEOUT = env_integer("RAY_SYMMETRIC_RUN_CLUSTER_WAIT_TIMEOUT", 30)
+
+
+def check_ray_already_started() -> bool:
+    import ray._private.services as services
+
+    # Try auto-detecting the Ray instance.
+    running_gcs_addresses = services.find_gcs_addresses()
+    return len(running_gcs_addresses) > 0
+
+
+def check_cluster_ready(nnodes, timeout=CLUSTER_WAIT_TIMEOUT):
+    """Wait for all nodes to start.
+
+    Raises an exception if the nodes don't start in time.
+    """
+    start_time = time.time()
+    current_nodes = 1
+    ray.init(ignore_reinit_error=True)
+
+    while time.time() - start_time < timeout:
+        time.sleep(5)
+        current_nodes = len(ray.nodes())
+        if current_nodes == nnodes:
+            return True
+        else:
+            click.echo(
+                f"Waiting for nodes to start... {current_nodes}/{nnodes} nodes started"
+            )
+    return False
+
+
+def check_head_node_ready(address: str, timeout=CLUSTER_WAIT_TIMEOUT):
+    start_time = time.time()
+    gcs_client = GcsClient(address=address)
+    while time.time() - start_time < timeout:
+        if gcs_client.check_alive([], timeout=1):
+            click.echo("Ray cluster is ready!")
+            return True
+        time.sleep(5)
+    return False
+
+
+def curate_and_validate_ray_start_args(run_and_start_args: List[str]) -> List[str]:
+    # Reparse the arguments to remove symmetric_run arguments.
+    ctx = symmetric_run.make_context("_", run_and_start_args, resilient_parsing=True)
+    cleaned_args = list(ctx.params["ray_args_and_entrypoint"])
```

Comment:
nice.

Why passing `_` as info_name to `make_context`?

Reply:
mainly because i'm just abusing this to parse args rather than actually have this context be user-facing